obspy/core/event/event.py

```diff
@@ -311,6 +319,23 @@ def write(self, filename, format, **kwargs):
         from .catalog import Catalog
         Catalog(events=[self]).write(filename, format, **kwargs)
 
+    def scope_resource_ids(self):
+        """
+        Ensure all resource_ids in event instance are event-scoped.
+
+        This will ensure the resource_ids refer to objects in the event
+        structure when possible.
+        """
+        gen = _yield_resource_id_parent_attr(self)
+
+        for resource_id, parent, attr in gen:
+            if attr == 'resource_id':
+                resource_id.set_referred_object(parent, parent=self,
+                                                warn=False)
+            else:
+                resource_id._parent_key = self
+                resource_id._object_id = None
```

Comment:
How are things like the derived origin ids scoped after this function has been called? `_parent_key` would be set but `_parent_id_tree` and things like this might not be set. But I just be misunderstanding some things.

Reply:
Any object that is referenced by a resource_id should have the attribute `resource_id`. Lines 332-334 take care of this case, which would involve logging the identities of all the objects with `resource_id` attributes into the proper place in the `_parent_id_tree`. Then, resource ids that refer to other objects in the same event (but not their parent) such as the `pick_id` of the `Arrival` object will find the correct object when `get_referred_object` is called because it was previously put into `_parent_id_tree`. Because `_object_id` on has been set to `None` it will not raise a warning. I added another test to specifically check for warnings being raised in this case.

Reply:
:+1: Thanks for the explanation. After spending some more time I think I now fully understand your changes and how everything works. Nice job!